### PR TITLE
feat: Align chectl telemetry with redhat devtools

### PR DESCRIPTION
**Pull request by flacatus**
Signed-off-by: Flavius Lacatusu <flacatus@redhat.com>

### What does this PR do?
Reff issue: https://github.com/eclipse/che/issues/18861

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [x] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [ ] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [ ] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [ ] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [ ] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [ ] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [ ] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### CODE REVIEW

**Comment by flacatus** [comment]
Woopra metrics related to PR

image

Also I've tested the PR with CRC and both have the same anonymousId when send events to segment

File: src/hooks/analytics/segment-adapter.ts [hunk 1]
```diff
@@ -40,6 +47,37 @@ export class SegmentAdapter {
...
+
+    if (fs.existsSync(anonymousIdPath)) {
+      return fs.readFileSync(anonymousIdPath, 'utf8')
+    } else {
+      const anonymousId = v4()
+      if (!fs.existsSync(anonymousIdPath)) {
+        fs.mkdirSync(path.join(os.homedir(), '.redhat'))
+      }
+
+      fs.writeFileSync(anonymousIdPath, anonymousId, { encoding: 'utf8' })
```
Comment (tolusha) [inline src/hooks/analytics/segment-adapter.ts:64]: We should catch error here.

File: src/hooks/analytics/segment-adapter.ts [hunk 2]
```diff
@@ -40,6 +47,37 @@ export class SegmentAdapter {
...
+   */
+  static getAnonymousId(): string {
+    const anonymousIdPath = path.join(os.homedir(), '.redhat', 'anonymousId')
+
+    if (fs.existsSync(anonymousIdPath)) {
+      return fs.readFileSync(anonymousIdPath, 'utf8')
+    } else {
+      const anonymousId = v4()
+      if (!fs.existsSync(anonymousIdPath)) {
+        fs.mkdirSync(path.join(os.homedir(), '.redhat'))
```
Comment (tolusha) [inline src/hooks/analytics/segment-adapter.ts:61]: We should catch error here.

File: src/util.ts [hunk 3]
```diff
@@ -272,3 +275,22 @@ export async function findWorkingNamespace(flags: any): Promise<string> {
...
+
+export function getPlatform(): string {
+  const platform: string = os.platform()
+  if (platform.startsWith('win')) {
+    return 'Windows'
+  }
+  if (platform.startsWith('darwin')) {
+    return 'Mac'
+  }
+  return platform.charAt(0).toUpperCase() + platform.slice(1)
```
Comment (tolusha) [inline src/util.ts:287]: That's not obvious.

File: src/util.ts [hunk 4]
```diff
@@ -272,3 +275,22 @@ export async function findWorkingNamespace(flags: any): Promise<string> {
 
   return DEFAULT_CHE_NAMESPACE
 }
+
+export function getPlatform(): string {
+  const platform: string = os.platform()
+  if (platform.startsWith('win')) {
```
Comment (tolusha) [inline src/util.ts:281]: Do we really need to convert `win` to `Windows`, `darwin` to `Mac`?

Comment (flacatus) [reply src/util.ts:281]: Is not necessary. I can remove it

File: src/hooks/analytics/segment-adapter.ts (hunk 2 above)
Comment (flacatus) [reply src/hooks/analytics/segment-adapter.ts:61]: ok
@tolusha  In case of catch(e) {} Should we print a message there or don't nothing?

**Review by AndrienkoAleksandr — APPROVED** [review]

File: src/hooks/analytics/segment-adapter.ts [hunk 5]
```diff
@@ -40,6 +48,41 @@ export class SegmentAdapter {
...
+        }
+
+        fs.writeFileSync(anonymousIdPath, anonymousId, { encoding: 'utf8' })
+
+        return anonymousId
+      }
+    } catch (error) {
+      cli.debug(`Failed to store anonymousId ${error}`)
+    }
+  }
```
Comment (tolusha) [inline src/hooks/analytics/segment-adapter.ts:73]:
```suggestion
  static getAnonymousId(): string | undefined {
    const anonymousIdPath = path.join(os.homedir(), '.redhat', 'anonymousId')
    let anonymousId = v4()
    try {
      if (fs.existsSync(anonymousIdPath)) {
        anonymousId = fs.readFileSync(anonymousIdPath, 'utf8')
      } else {
        if (!fs.existsSync(anonymousIdPath)) {
          fs.mkdirSync(path.join(os.homedir(), '.redhat'))
        }

        fs.writeFileSync(anonymousIdPath, anonymousId, { encoding: 'utf8' })
      }
    } catch (error) {
      cli.debug(`Failed to store anonymousId ${error}`)
    }
    
    return anonymousId.trim()
  }
```

File: src/hooks/analytics/segment-adapter.ts [hunk 6]
```diff
@@ -49,14 +92,51 @@ export class SegmentAdapter {
     this.segment.track({
       anonymousId: this.id,
       event: options.command.replace(':', ' '),
+
+      context: await this.getSegmentEventContext(),
       properties: {
         ...pick(options.flags, ['platform', 'installer']),
-        command: options.command
+        command: options.command,
+        version: getProjectVersion()
```
Comment (tolusha) [inline src/hooks/analytics/segment-adapter.ts:100]: It seems we collection version twice.
Is it ok to remove it from this place?

File: src/util.ts [hunk 7]
```diff
@@ -272,3 +275,11 @@ export async function findWorkingNamespace(flags: any): Promise<string> {
 
   return DEFAULT_CHE_NAMESPACE
 }
+
+export async function getDistribution(): Promise<string | undefined> {
+  if (os.platform() === 'linux') {
+    const platorm = await promisify(getos)() as getos.LinuxOs
```
Comment (tolusha) [inline src/util.ts:281]: Is it possible to get some errors here?

File: src/hooks/analytics/segment-adapter.ts (hunk 6 above)
Comment (flacatus) [reply src/hooks/analytics/segment-adapter.ts:100]: Yes,  Woopra ignore version from segmentEventContext. I will removed from here: https://github.com/che-incubator/chectl/blob/red_hat_tele/src/hooks/analytics/segment-adapter.ts#L134

**Review by tolusha — APPROVED** [review]

**Pull request merged**